rat/compiler2.py

```diff
@@ -0,0 +1,737 @@
+        #                     # If variable is known to have subscripts then check that
+        #                     # the number of subscripts are compatible with existing
+        #                     # ones
+        #                     if len(variable.subscripts) > 0:
+        #                         if len(subscript_names) != len(variable.subscripts):
+        #                             msg = f"{len(subscript_names)} found, previously used with {len(variable.subscripts)}"
+        #                             raise CompileError(msg, node)
+
+        #                     # Fold all the shift expressions to simple integers
+        #                     # TODO: Is this necessary?
+        #                     # shifts, pad_needed = self._get_shifts_and_padding(symbol.subscript)
+        #                     # variable.pad_needed = pad_needed
+
+        #                     # Extra checks for secondary variables
+        #                     # TODO: Is this necessary?
+        #                     # if symbol_key != primary_symbol_key:
+        #                     # TODO: This should probably be supported in the future
+        #                     # right now I'm leaving it off because I'm not quite sure
+        #                     # what the behavior should be and I don't have an example
+        #                     # model in mind (my guess is access should give zeros)
+        #                     #     if any(shift != 0 for shift in shifts):
+        #                     #         msg = f"Shifted access on a secondary parameter is not allowed"
+        #                     #         raise CompileError(msg, symbol.range)
+
+        #     walker = DataframeCompatibilityWalker(primary_ast_variable.name, data_names, self.variable_table)
+        #     walker.walk(statement)
+
+        # Trace the program to determine parameter domains and check data domains
+        for _ in range(self.max_trace_iterations):
+            # Reset all the parameter tracers
+            tracers = {}
+            for name in self.variable_table:
+                variable = self.variable_table[name]
+                if not variable.tracer:
+                    variable.tracer = Tracer()
+                tracers[name] = variable.tracer.copy()
+
+            for statement_info in self.statements:
+                # Find the primary variable
+                primary_ast_variable = statement_info.primary
+                primary_variable = self.variable_table[primary_ast_variable.name]
+
+                for row in primary_variable.itertuples():
+                    executor = TraceExecutor(tracers, row._asdict())
+                    executor.walk(statement_info.statement)
+
+            found_new_traces = False
+            for variable_name, tracer in tracers.items():
+                variable = self.variable_table[variable_name]
+                found_new_traces |= variable.ingest_new_trace(tracer)
```

Comment:
Hmm. Updating the variable table at the end results in multiple tracing iterations needed to evaluate non-recursive programs. Updating the variable table while we're tracing will fix this, since then multiple dependent variables can be traced in a single trace, but then you'd need more logic and code to check if there are new traces(comparing the lengths of all tracer arguments before and after the trace iteration). Would this be worth?

Reply:
Until we see a performance problem I lean no. I expect we'll see a performance problem eventually lol.